src/libpgagroal/utils.c

```diff
@@ -954,128 +955,124 @@ pgagroal_backtrace(void)
+      if (strncmp(command, command_table[i].command, MISC_LENGTH) == 0)
+      {
+         command_match[command_match_count++] = i;
+      }
+   }
+   if (command_match_count == 0)
+   {
+      sprintf(cmd->error_message, "Invalid command '%s'\n", command);
       return false;
    }
 
-   if (subcommand)
+   /* Parse subcommand, and exit if there is no match */
+   CONSUME(subcommand);
+   subcommand = subcommand ? subcommand : "";
+   for (int i = 0; i < command_match_count; i++)
    {
-      // thre must be a subcommand check
-      offset++;
-
-      if (argc <= offset)
-      {
-         // not enough command args!
-         return false;
-      }
-
-      if (strncmp(argv[offset], subcommand, MISC_LENGTH))
+      if (strncmp(subcommand, command_table[command_match[i]].subcommand, MISC_LENGTH) == 0)
       {
-         return false;
+         cmd->command_code = command_match[i]; /* pushes down the matches */
       }
    }
-
-   if (key)
+   if (cmd->command_code == COMMAND_UNKNOWN)
    {
-      // need to evaluate the database or server or configuration key
-      offset++;
-      *key = argc > offset ? argv[offset] : default_key;
-      if (*key == NULL || strlen(*key) == 0)
+      if (EMPTY_STR(subcommand))
       {
-         goto error;
+         sprintf(cmd->error_message, "Command '%s' requires a subcommand\n", command);
+         return false;
       }
-
-      // do I need also a value?
-      if (value)
+      else
       {
-         offset++;
-         *value = argc > offset ? argv[offset] : default_value;
-
-         if (*value == NULL || strlen(*value) == 0)
+         offset--;
+         subcommand = "";
+         for (int i = 0 ; i < command_match_count; i++)
          {
-            goto error;
+            if (strncmp(subcommand, command_table[command_match[i]].subcommand, MISC_LENGTH) == 0)
+            {
+               cmd->command_code = command_match[i]; /* pushes down the matches */
+            }
+         }
+         if (cmd->command_code == COMMAND_UNKNOWN)
+         {
+            sprintf(cmd->error_message, "Invalid subcommand '%s' for command '%s'\n", subcommand, command);
+            return false;
          }
-
       }
    }
 
-   return true;
-
-error:
-   return false;
-}
-
-bool
-parse_deprecated_command(int argc,
-                         char** argv,
-                         int offset,
-                         char* command,
-                         char** value,
-                         char* deprecated_by,
-                         unsigned int deprecated_since_major,
-                         unsigned int deprecated_since_minor)
-{
-   // sanity check: if no arguments, nothing to parse!
-   if (argc <= offset)
+   /* Iterate until finds an accepted_arg_count that is equal or greater than the typed command arg_count */
+   cmd->arg_count = argc - offset;
+   int j;
+   for (j = 0; j < MISC_LENGTH && command_table[cmd->command_code].accepted_argument_count[j] < cmd->arg_count; j++)
+   {
+      ;
```

Comment:
I think a `for` with an `if ... break` would be a little more readable than this.